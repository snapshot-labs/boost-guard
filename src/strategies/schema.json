{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$ref": "#/definitions/Strategy",
  "definitions": {
    "Strategy": {
      "type": "object",
      "properties": {
        "strategy": {
          "type": "string",
          "enum": ["snapshot"]
        },
        "version": {
          "type": "number",
          "enum": [1]
        },
        "params": {
          "type": "object"
        }
      },
      "required": ["strategy", "params"],
      "additionalProperties": false
    }
  }
}
