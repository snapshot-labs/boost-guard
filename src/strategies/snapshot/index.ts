import fetch from 'cross-fetch';
import { BigNumber } from '@ethersproject/bignumber';

const SNAPSHOT_HUB_URL = 'https://hub.snapshot.org/graphql';
const SNAPSHOT_TESTNET_HUB_URL = 'https://testnet.hub.snapshot.org/graphql';

export default async function strategy(recipient: string, params: any): Promise<string> {
  const url = params.env === 'testnet' ? SNAPSHOT_TESTNET_HUB_URL : SNAPSHOT_HUB_URL;

  const init = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      query: `
        query Vote($voter: String!, $proposal: String!) {
          votes(
            first: 1
            where: {
              voter: $voter
              proposal: $proposal
            }
          ) {
            vp
            vp_state
          }
        }
      `,
      variables: {
        voter: recipient,
        proposal: params.proposal
      }
    })
  };
  const res = await fetch(url, init);
  const { data } = await res.json();

  if (!data.votes[0] || data.votes[0].vp === 0) return '0';

  if (params.type === 'ratio')
    return BigNumber.from(params.amount).mul(BigNumber.from(data.votes[0].vp)).toString();

  if (params.type === 'fixed') return params.amount;

  return '0';
}
