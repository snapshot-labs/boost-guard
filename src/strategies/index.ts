import snapshot from './snapshot';

export default {
  snapshot
};
