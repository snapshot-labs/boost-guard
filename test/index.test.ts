import strategies from '../src/strategies';

describe('Strategies', () => {
  it('snapshot', async () => {
    const recipient = '0xeF8305E140ac520225DAf050e2f71d5fBcC543e7';
    const params = {
      proposal: '0x0021dc765768342d269184cb46e1cf17e6609559973d88bd01a292a4f390caa4',
      type: 'fixed',
      amount: '1000'
    };

    const status = await strategies.snapshot(recipient, params);

    expect(status).toMatchSnapshot();
  });
});
