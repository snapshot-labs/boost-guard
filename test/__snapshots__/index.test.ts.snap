// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Strategies snapshot 1`] = `"1000"`;
